File: 2020/Java/src/aoc/InputReader.java
package aoc;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;

public abstract class InputReader {

	/**
	 * Gets the input file to read for the given day.
	 * 
	 * @param day the day for which to get the input file.
	 * @return the file to read for this days puzzle.
	 */
	public static File getInputFile(int day) {
		File input = new File("input").getAbsoluteFile();
		if (!input.exists() || !input.isFile()) {
			System.out.println(input.getPath() + " does not exist!");
			input = new File(input.getPath() + ".txt");
			System.out.println("trying " + input.getPath() + " instead.");
		}

		if (!input.exists() || !input.isFile()) {
			System.out.println(input.getPath() + " does not exist!");
			input = new File(input.getParentFile().getParent(), "input");
			input = new File(input, "Day" + day + ".txt");
			System.out.println("trying " + input.getPath() + " instead.");
		}

		if (!input.exists() || !input.isFile()) {
			System.err.println("None of the expected inputs exist!");
			System.exit(1);
		} else {
			System.out.println("Using input file " + input.getPath() + ".");
		}

		return input;
	}

	/**
	 * Reads all the lines from the input file as a list of strings.
	 * 
	 * @param day the day for which to read the input file.
	 * @return an {@link ArrayList} containing a string for each line of the input
	 *         file.
	 * @throws IOException if an I/O error occurs reading from the file or a
	 *                     malformed or unmappable byte sequence is read.
	 */
	public static List<String> readInputFileLines(int day) throws IOException {
		List<String> lines = new ArrayList<>();
		Files.readAllLines(getInputFile(day).toPath()).forEach(line -> lines.add(line.trim()));
		return lines;
	}

	/**
	 * Reads all the lines from the input file as strings and parses them to longs.
	 * 
	 * @param day the day for which to read the input file.
	 * @return an {@link ArrayList} containing a long for each line of the input
	 *         file.
	 * @throws IOException if an I/O error occurs reading from the file or a
	 *                     malformed or unmappable byte sequence is read.
	 */
	public static List<Long> readInputFileNumbers(int day) throws IOException {
		List<String> lines = readInputFileLines(day);
		List<Long> numbers = new ArrayList<>();
		lines.forEach(line -> numbers.add(Long.parseLong(line)));
		return numbers;
	}

}

File: 2020/Python/src/InputReader.py
import os
import os.path as path
import sys


def getInputFile(day):
    """Gets the input file to read for the given day.
    
    Parameters
    ----------
    day: int
        The day for which to read the input file.
    
    Returns
    -------
    str
        The path of the file to read for this days puzzle.
    """
    inf = path.dirname(os.getcwd())
    inf = path.join(inf, "input")
    if not path.exists(inf) or not path.isfile(inf):
        print(inf + " does not exist!")
        inf += ".txt"
        print("trying %s instead." % inf)

    if not path.exists(inf) or not path.isfile(inf):
        print(inf + " does not exist!")
        inf = path.dirname(path.dirname(path.dirname(inf)))
        inf = path.join(inf, "input")
        inf = path.join(inf, "Day%d.txt" % day)
        print("trying %s instead." % inf)

    if not path.exists(inf) or not path.isfile(inf):
        print("None of the expected inputs exist!", file=sys.stderr)
        sys.exit(1)
    else:
        print("Using input file %s." % inf)

    return inf


def readInputFileLines(day):
    """Reads all the lines from the input file as a list of strings.
    
    Parameters
    ----------
    day: int
        The day for which to read the input file.
    
    Returns
    -------
    list of str
        A list containing a string for each line of the input file.
    """
    inf = getInputFile(day)
    lines = []
    inputFile = open(inf, 'r')
    for line in inputFile.readlines():
        lines.append(line.strip())

    inputFile.close()
    return lines


def readInputFileNumbers(day):
    """Reads all the lines from the input file as strings and parses them to ints.
    
    Parameters
    ----------
    day: int
        The day for which to read the input file.
    
    Returns
    -------
    list of int
        A list containing an int for each line of the input file.
    """
    numbers = []
    for line in readInputFileLines(day):
        numbers.append(int(line))

    return numbers
